internal/coreinternal/pdatautil/map.go

```diff
@@ -0,0 +1,104 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pdatautil // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil"
+
+import (
+	"encoding/binary"
+	"hash"
+	"math"
+	"sort"
+
+	"github.com/cespare/xxhash/v2"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+var (
+	valEmpty        = []byte{'\xf5'}
+	valBytesPrefix  = []byte{'\xf6'}
+	valStrPrefix    = []byte{'\xf7'}
+	valBoolTrue     = []byte{'\xf8'}
+	valBoolFalse    = []byte{'\xf9'}
+	valIntPrefix    = []byte{'\xfa'}
+	valDoublePrefix = []byte{'\xfb'}
+	valMapPrefix    = []byte{'\xfc'}
+	valMapSuffix    = []byte{'\xfd'}
+	valSlicePrefix  = []byte{'\xfe'}
+	valSliceSuffix  = []byte{'\xff'}
+)
+
+// MapHash return a hash for the provided map.
+// Maps with the same underlying key/value pairs in different order produce the same deterministic hash value.
+func MapHash(m pcommon.Map) uint64 {
+	h := xxhash.New()
+	writeMapHash(h, m)
+	return h.Sum64()
+}
+
+func writeMapHash(h hash.Hash, m pcommon.Map) {
```

Comment:
Would like to see a benchmark of this implementation for a classic case like 4/8/16 arguments

Reply:
Added benchmarks:

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems
BenchmarkMapHashFourItems-10                  	 3161750	       377.5 ns/op	     272 B/op	      10 allocs/op
BenchmarkMapHashEightItems
BenchmarkMapHashEightItems-10                 	 1552203	       782.6 ns/op	     440 B/op	      18 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	  929382	      1228 ns/op	     768 B/op	      31 allocs/op
PASS
```

I'll also try SHA1 to generate uuidv5 to see the difference

Reply:
SHA1 implementation just a bit slower:

```golang
func MapHash(m pcommon.Map) [sha1.Size]byte {
	h := sha1.New()
	writeMapHash(h, m)
	b := h.Sum(nil)
	res := [sha1.Size]byte{}
	copy(res[:], b)
	return res
}
```

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/pdatautil
BenchmarkMapHashFourItems
BenchmarkMapHashFourItems-10                  	 2616472	       452.1 ns/op	     328 B/op	      11 allocs/op
BenchmarkMapHashEightItems
BenchmarkMapHashEightItems-10                 	 1315567	       907.3 ns/op	     496 B/op	      19 allocs/op
BenchmarkMapHashWithEmbeddedSliceAndMap
BenchmarkMapHashWithEmbeddedSliceAndMap-10    	  795008	      1496 ns/op	     824 B/op	      32 allocs/op
PASS
```

Reply:
Do you think we should switch to sha1/uuid5?